File: clients/ruby/test/routes_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "json"
require "novamem"

class RoutesTest < Minitest::Test
  # proved by: renaming Novamem::Client#session_recap fails this test.
  def test_every_route_is_accounted
    routes = JSON.parse(File.read(File.expand_path("../../contract/routes.json", __dir__)))
    routes.each do |key, r|
      Array(r["methods"]).each do |m|
        cls, meth = m["name"].split(".")
        name = meth.gsub(/([A-Z])/) { "_#{::Regexp.last_match(1).downcase}" }.sub(/\A_/, "")
        assert Novamem.const_get(cls).method_defined?(name), "#{key} #{cls}##{name}"
      end
    end
  end
end

File: clients/typescript/test/routes.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import * as nm from "../src/index.js";

const ROUTES = JSON.parse(
  readFileSync(
    new URL("../../../contract/routes.json", import.meta.url),
    "utf8"
  )
);
const camel = (s: string) => s[0].toLowerCase() + s.slice(1);

// proved by: renaming Client.prototype.sessionRecap fails this test.
test("every routes.json method resolves", () => {
  for (const [key, r] of Object.entries<any>(ROUTES)) {
    for (const m of r.methods ?? []) {
      const [cls, meth] = m.name.split(".");
      assert.equal(
        typeof (nm as any)[cls].prototype[camel(meth)],
        "function",
        `${key} ${m.name}`
      );
    }
  }
});
